src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,48 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static final int DEFAULT_STACK_SIZE = 10;
+    private T[] stackArray;
+    private int size;
 
+    public MateStack() {
+        stackArray = (T[]) new Object[DEFAULT_STACK_SIZE];
+    }
+
+    public void push(T value) {
+        if (size == stackArray.length) {
+            resize();
+        }
+        stackArray[size++] = value;
     }
 
     public T peek() {
-        return null;
+        checkSize();
+        return stackArray[size - 1];
     }
 
     public T pop() {
-        return null;
+        checkSize();
+        T removedObject = stackArray[--size];
+        stackArray[size] = null;
+        return removedObject;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void checkSize() {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+    }
+
+    private void resize() {
+        T[] newArray = (T[]) new Object[stackArray.length * 3 / 2 + 1];
```

Comment:
and do we even need to grow the array when size is 0 or 1, considering default capacity is 10 and we call this method when `capacity == size`?

Reply:
Yes, because we can change the default capacity to 0 or 1. Or implement a trim method that will reduce the size of the array after each pop() call.